File: Tutorial5-3aDagger-Android/src/main/java/com/example/tutorial5_1dagger_android/di/ApplicationComponent.java
package com.example.tutorial5_1dagger_android.di;


import android.app.Application;

import com.example.tutorial5_1dagger_android.MyApplication;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjector;
import dagger.android.support.AndroidSupportInjectionModule;

// This is parent component
@Component(modules = {ApplicationModule.class, AndroidSupportInjectionModule.class})
@Singleton
public interface ApplicationComponent extends AndroidInjector<MyApplication> {
    @Component.Builder
    interface Builder {
        @BindsInstance
        Builder application(Application application);

        ApplicationComponent build();
    }
}

File: Tutorial4-6SubComponentBuilders-MultipleModulesAndActvities/src/main/java/com/example/tutorial4_6subcomponentbuilders_multiplemodulesandactvities/di/ApplicationComponent.kt
package com.example.tutorial4_6subcomponentbuilders_multiplemodulesandactvities.di

import android.app.Application
import com.example.tutorial4_6subcomponentbuilders_multiplemodulesandactvities.di.scope.SubComponentsModule
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

// This is parent component
@Component(modules = [ApplicationModule::class, SubComponentsModule::class])
@Singleton
interface ApplicationComponent {

    /**
     * Calls SubComponent Builder in SecondActivity and MyFragment using ApplicationComponent
     */
    fun toastMakerBuilder(): ToastAndPreferencesSubComponent.Builder

    fun dummyDependencyBuilder(): DummyDependencySubComponent.Builder

    @Component.Builder
    interface Builder {

        @BindsInstance
        fun application(application: Application): Builder

        fun build(): ApplicationComponent
    }
}